// @colyseus/core/build/router/node.mjs
// packages/core/src/router/node.ts
import { URL } from "url";
import * as matchMaker from "../MatchMaker.mjs";
import { setResponse } from "@colyseus/better-call/node";
import { postMatchmakeMethod } from "./default_routes.mjs";
function readBody(req) {
  return new Promise((resolve, reject) => {
    let data = "";
    req.on("data", (chunk) => {
      data += chunk.toString();
    });
    req.on("end", () => resolve(data ? JSON.parse(data) : {}));
    req.on("error", reject);
  });
}
function prereadRequestBodies(server) {
  const listeners = server.listeners("request");
  const run = (req, res) => {
    for (const listener of listeners) {
      listener.call(server, req, res);
    }
  };
  server.removeAllListeners("request");
  server.on("request", (req, res) => {
    const method = req.method ?? "GET";
    const needsBody = method !== "GET" && method !== "HEAD" && method !== "OPTIONS" && req.body === void 0 && Number(req.headers["content-length"]) > 0;
    if (!needsBody) {
      run(req, res);
      return;
    }
    const chunks = [];
    req.on("data", (chunk) => chunks.push(chunk));
    req.on("end", () => {
      const raw = Buffer.concat(chunks).toString("utf8");
      if ((req.headers["content-type"] || "").includes("application/json")) {
        try {
          req.body = JSON.parse(raw || "{}");
        } catch {
          req.body = raw;
        }
      } else {
        req.body = raw;
      }
      run(req, res);
    });
    req.on("error", () => run(req, res));
  });
}
function getCorsHeaders(req, headers) {
  return {
    ...matchMaker.controller.DEFAULT_CORS_HEADERS,
    ...matchMaker.controller.getCorsHeaders(headers)
  };
}
function createNodeMatchmakingMiddleware() {
  return async (req, res, next) => {
    const url = new URL(req.url || "/", "http://localhost");
    const isMatchmakeRoute = url.pathname.startsWith(`/${matchMaker.controller.matchmakeRoute}/`);
    if (!isMatchmakeRoute) {
      next();
      return;
    }
    const headers = new Headers(req.headers);
    const corsHeaders = getCorsHeaders(req, headers);
    if (req.method === "OPTIONS") {
      res.writeHead(204, corsHeaders);
      res.end();
      return;
    }
    if (req.method !== "POST") {
      next();
      return;
    }
    const match = url.pathname.match(/^\/matchmake\/(\w+)\/(.+)/);
    if (!match) {
      next();
      return;
    }
    const [, method, roomName] = match;
    try {
      const response = await postMatchmakeMethod({
        params: { method, roomName },
        body: await readBody(req),
        headers: req.headers,
        request: { headers },
        asResponse: true
      });
      await setResponse(res, response);
    } catch {
      next();
    }
  };
}
export {
  createNodeMatchmakingMiddleware,
  prereadRequestBodies
};
